src/vumi2/transports/smpp/client.py

```diff
@@ -211,28 +211,29 @@ async def send_pdu(
         return response
 
     async def send_vumi_message(self, message: Message):
-        # At the moment this will always only return a single PDU. When we implement
-        # multipart messages, this error handling will need to be improved
+        # If we encounter an error in one of the segments of a multipart message,
+        # immediately send a nack and stop trying with the rest
         for pdu in await self.submit_sm_processor.handle_outbound_message(message):
             response = await self.send_pdu(pdu, check_response=False)
 
-        # Requests always get responses
-        response = cast(PDUResponse, response)
-
-        if response.status == CommandStatus.ESME_ROK:
-            event = Event(
-                user_message_id=message.message_id,
-                sent_message_id=message.message_id,
-                event_type=EventType.ACK,
-            )
-        else:
-            status_code = command_status_name_map[response.status.name]
-            nack_reason = command_status_value_map[status_code]["description"]
-            event = Event(
-                user_message_id=message.message_id,
-                sent_message_id=message.message_id,
-                event_type=EventType.NACK,
-                nack_reason=nack_reason,
-            )
+            # We will always get a response from a request
+            response = cast(PDU, response)
```

Comment:
Is there value in casting to something more specific than this?

Reply:
Maybe for autocomplete? Although the library isn't really great with that, even if you choose a specific PDU type, it's just the same base PDU fields, and then everything is in a dict-like params, so you don't get much. But we could do a `PDUResponse`, or since we know that we're always doing a `submit_sm`, maybe a `SubmitSMResp`, but since the code that creates the PDUs is separate to this, I'm worried that if in future it gets changed to give other kinds of PDUs, that we forget to change this.

What I really want is for `send_pdu` to have a dynamic type. If it's a request, then it returns a response. If it's a response, then it returns None. I tried to do this with type overloads, but it gives an error that the second overload will never match, and I can't understand why.

I think maybe I'll change that to a response, instead of the base PDU